common/workunit/workunit.cpp

```diff
@@ -5060,6 +5062,171 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+            wuTree->addPropTree("GraphProgress", graphProgressTree.getClear());
+        }
+    public:
+        CImportWorkUnitHelper(const char *_zapReportFileName, const char *_user)
+            : zapReportFileName(_zapReportFileName), user(_user) { };
+
+        IPTree *queryWUPTree()
+        {
+            return wuTree;
+        }
+        void setUNZIPDir(const IPropertyTree *directories, const char *component, const char *instance)
+        {   //Set a unique unzip folder inside the component's data folder
+            getConfigurationDirectory(directories, "data", component, instance, unzipDir);
+            unzipDir.append(PATHSEPSTR).append(zapReportFileName.length() - 4, zapReportFileName);
+            if (!user.isEmpty())
+                unzipDir.append("_").append(user);
+
+            CDateTime dt;
+            dt.setNow();
+            dt.getString(unzipDateTime);
+            unzipDir.append("_").append(dt.getTimeStamp());
+        }
+        int unzipZAPReport(const char *zapReportFilePath, const char *zapReportPassword)
+        {
+            Owned<IFile> unzipFolder = createIFile(unzipDir);
+            if (!unzipFolder->exists())
+                unzipFolder->createDirectory();
+            else
+            {//This should not happen. Just in case.
+                Owned<IDirectoryIterator> iter = unzipFolder->directoryFiles(nullptr, false, false);
+                ForEach(*iter)
+                {
+                    OwnedIFile thisFile = createIFile(iter->query().queryFilename());
+                    if (thisFile->isFile() == foundYes)
+                        thisFile->remove();
+                }
+            }
+
+            //Unzip ZAP Report
+            StringBuffer zipCommand("unzip ");
+            if (!isEmptyString(zapReportFilePath))
+                zipCommand.append(zapReportFilePath);
+            if (isEmptyString(zapReportPassword))
+                zipCommand.appendf("%s -d %s", zapReportFileName.get(), unzipDir.str());
+            else
+                zipCommand.appendf("%s -P %s -d %s", zapReportFileName.get(), zapReportPassword, unzipDir.str());
+            return (system(zipCommand.str()));
+        }
+        void buildWUPTreesFromZAPReport()
+        {
+            readWUXMLFileToPTree();
+            wuid.set(wuTree->queryName());
+            setImportDebugAttribute();
+
+            StringBuffer localIP, unzipDirWithIP("//");
+            IpAddress ipaddr = queryHostIP();
+            ipaddr.getIpText(localIP);
+            unzipDirWithIP.append(localIP).append(unzipDir).append(PATHSEPSTR);
+
+            //update log entries in WU XML;
+            updateWUProcessLogAttrs("Process/EclAgent/*", unzipDirWithIP);
+            updateWUProcessLogAttrs("Process/Thor/*", unzipDirWithIP);
+
+            //update QueryAssociatedFiles in WU XML;
+            updateWUQueryAssociatedFilesAttrs(localIP);
+
+            readGraphProgressFileToWUPTree();
```

Comment:
it's a little inefficient to load the file and set it into the wuTree.
For it later to be removed (pruned) then added separately to /GraphProgress, when really it was separate in the 1st place.

I think this would better as a separate step. (but still within CWorkUnitFactory::importWorkUnit) after the creation of the wuid and import of it's xml (see comments above), then add progress tree.

e.g. something like:
```
    CImportWorkUnitHelper helper(zapReportFileName, user);
    Owned<IWorkUnit> newWuid = createWorkUnit(app, user, secMgr, secUser);
    helper.setUNZIPDir(directories, component, instance, newWuid->queryWuid());
    helper.unzipZAPReport(zapReportFilePath, zapReportPassword);   
    helper.buildWUPTreesFromZAPReport(); // keeps progress tree separately

    Owned<IPropertyTree> wuidTree = helper.getWorkUnitTree();
    newWuid->import(wuidTree);
    Owned<IPropertyTree> progressTree = helper.getPropgressTree();
    importProgress(newWuid->queryWuid(), progressTree);
```

Reply:
@jakesmith In fact, I DO like this approach better because it is much easy to understand and to implement. The existing approach is based on your comment: 'I would be better/more efficient to do it in one call. I would add a new createWorkUnit method that takes a IPropertyTree. Load the archive workunit xml to a ptree, patch up where necessary, than add via the new createWorkUnit, that copies the content into the newly created uniquely named wuid.' As your comment said, the existing approach has only one call to Dali WUs Tree. Is the existing approach more efficient than this approach?

Reply:
@wangkx - efficiency is important, whilst not a critical area, it could be time consuming and is easy to avoid adding then removing the progress tree here.
a new createWorkUnit method (as opposed to altering the existing) would have worked too, if you had passed the workunit tree and progress tree separately.
The issue here is you add it to a loaded workunit xml , but then prune it later, before making the new connection for the GraphProgress.

The both new and approach, must make 2 connections - one for the /WorkUnit/... and one for the GraphProgress, there is no difference in efficiency wrt as far as that is concerned, in both cases it has to copy 2 trees. So as far a Dali traffic is concerned there is no difference.
It's the local copying/pruning of the GraphProgress that can be avoided with this approach.

Reply:
ok